From 7160f8bf80d396b96524cbd57d7dcc6018ab54dc Mon Sep 17 00:00:00 2001
From: Chen Jun <chenjun102@huawei.com>
Date: Mon, 8 May 2023 11:22:59 +0000
Subject: [PATCH] iommu/arm-smmu-v3: Enable sva for somes non-stander pcie
 devices

We have some non-stander pcie devices without ecam.

To enable sva for those pcie devices, we need to set ssid_bits in
arm_smmu_probe_device, and need a new function to enable sva.

Signed-off-by: Chen Jun <chenjun102@huawei.com>
---
 drivers/iommu/Kconfig                         | 11 +++
 .../iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c   | 18 +++++
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c   | 72 +++++++++++++++++++
 3 files changed, 101 insertions(+)

diff --git a/drivers/iommu/Kconfig b/drivers/iommu/Kconfig
index 27a38a867..a9ce4bcb6 100644
--- a/drivers/iommu/Kconfig
+++ b/drivers/iommu/Kconfig
@@ -442,4 +442,15 @@ config AGENT_SMMU_ATOS
 	help
 	  provide a interface to translate iova to pa by agent smmu.
 
+config SMMU_ENFORCE_PASID
+	bool "SMMU pasid enforce for some specific devices"
+	depends on ARM_SMMU_V3=y
+	help
+	  Some non-stander pcie devices do not have ecam. If we want to enable sva for
+	  those devices, we should modify the ssid_bits.
+	  Using the smmu.enforce_pasid cmdline, to collect the device which ssid_bits
+	  should be modified to the value given.
+
+	  if unsure, say N here.
+
 endif # IOMMU_SUPPORT
diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c
index 91ba77a42..b15203cbe 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3-sva.c
@@ -537,6 +537,24 @@ static void arm_smmu_master_sva_disable_iopf(struct arm_smmu_master *master)
 	iopf_queue_remove_device(master->smmu->evtq.iopf, dev);
 }
 
+/*
+ * To enable sva for pcie devices with no ecam.
+ * The pcie devices can not enable sva becase iopf can not be enabled.
+ * Make sure no page fault casued by those devices.
+ */
+int arm_smmu_master_enable_sva_enforce(struct device *dev)
+{
+	struct arm_smmu_master *master = dev_iommu_priv_get(dev);
+
+	if (!master)
+		return -ENODEV;
+
+	master->sva_enabled = true;
+
+	return 0;
+}
+EXPORT_SYMBOL(arm_smmu_master_enable_sva_enforce);
+
 int arm_smmu_master_enable_sva(struct arm_smmu_master *master)
 {
 	int ret;
diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 7d907d140..60c716071 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -90,6 +90,73 @@ static int __init arm_smmu_bypass_dev_setup(char *str)
 __setup("smmu.bypassdev=", arm_smmu_bypass_dev_setup);
 #endif
 
+#ifdef CONFIG_SMMU_ENFORCE_PASID
+/*
+ * We have some non-standard pcie devices which has no ecam.
+ * The device's ssid_bits is set to 0 because no pasid cap
+ * be read from ecam during iommu_probe_device.
+ *
+ * ssid_bits is used to alloc cdtab and is checked in iommu_bind_device.
+ * If we want to enable sva for those devices, we need to modifiy
+ * ssid_bits.
+ */
+struct smmu_pasid_enforced_device {
+	unsigned short vendor;
+	unsigned short device;
+	unsigned short ssid_bits;
+};
+#define MAX_SMMU_PASID_ENFORCED_DEV 16
+
+static struct smmu_pasid_enforced_device smmu_pasid_enforced_devices[MAX_SMMU_PASID_ENFORCED_DEV];
+static int smmu_pasid_enforced_devices_num;
+
+static int __init arm_smmu_enforce_pasid_setup(char *str)
+{
+	unsigned short vendor;
+	unsigned short device;
+	unsigned short ssid_bits;
+	int ret;
+
+	ret = sscanf(str, "%hx:%hx:%hd", &vendor, &device, &ssid_bits);
+	if (ret != 3)
+		return -EINVAL;
+
+	if (smmu_pasid_enforced_devices_num >= MAX_SMMU_PASID_ENFORCED_DEV)
+		return -ERANGE;
+
+	smmu_pasid_enforced_devices[smmu_pasid_enforced_devices_num].vendor = vendor;
+	smmu_pasid_enforced_devices[smmu_pasid_enforced_devices_num].device = device;
+	smmu_pasid_enforced_devices[smmu_pasid_enforced_devices_num].ssid_bits = ssid_bits;
+	smmu_pasid_enforced_devices_num++;
+
+	return 0;
+}
+
+/* Usage: smmu.enforce_pasid=vender:device:ssid_bits */
+__setup("smmu.enforce_pasid=", arm_smmu_enforce_pasid_setup);
+
+static bool smmu_device_enforce_pasid(struct device *dev, unsigned int *ssid_bits)
+{
+	int i;
+	struct pci_dev *pdev;
+
+	if (!dev_is_pci(dev))
+		return false;
+
+	pdev = to_pci_dev(dev);
+
+	for (i = 0; i < smmu_pasid_enforced_devices_num; i++) {
+		if ((smmu_pasid_enforced_devices[i].vendor == pdev->vendor) &&
+		    (smmu_pasid_enforced_devices[i].device == pdev->device)) {
+			*ssid_bits = smmu_pasid_enforced_devices[i].ssid_bits;
+			return true;
+		}
+	}
+
+	return false;
+}
+#endif /* CONFIG_SMMU_ENFORCE_PASID */
+
 enum arm_smmu_msi_index {
 	EVTQ_MSI_INDEX,
 	GERROR_MSI_INDEX,
@@ -3384,6 +3451,11 @@ static struct iommu_device *arm_smmu_probe_device(struct device *dev)
 
 	arm_smmu_init_pri(master);
 
+#ifdef CONFIG_SMMU_ENFORCE_PASID
+	if (smmu_device_enforce_pasid(dev, &master->ssid_bits))
+		dev_info(dev, "enforce pasid to %d\n", master->ssid_bits);
+#endif /* CONFIG_SMMU_ENFORCE_PASID */
+
 	return &smmu->iommu;
 
 err_free_master:
-- 
2.17.1

